mesonbuild/modules/unstable_wayland.py

```
@@ -0,0 +1,121 @@
+            name, _ = os.path.splitext(base_name)
+
+            code = CustomTarget(
+                f'{name}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{scope}-code', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-protocol.c'],
+                backend=state.backend,
+            )
+            targets.append(code)
+
+            header = CustomTarget(
+                f'{name}-{side}-protocol',
+                state.subdir,
+                state.subproject,
+                [self.scanner_bin, f'{side}-header', '@INPUT@', '@OUTPUT@'],
+                [xml_file],
+                [f'{name}-{side}-protocol.h'],
+                backend=state.backend,
+            )
+            targets.append(header)
+
+        return ModuleReturnValue(targets, targets)
+
+    @typed_pos_args('wayland.find_protocol', str)
+    @typed_kwargs(
+        'wayland.find_protocol',
+        KwargInfo('state', str, default='stable', validator=in_set_validator({'stable', 'staging', 'unstable'})),
+        KwargInfo('version', (int, NoneType)),
+    )
+    def find_protocol(self, state, args, kwargs):
+        base_name = args[0]
+        xml_state = kwargs['state']
+        version = kwargs['version']
+
+        if (xml_state == 'staging' or xml_state == 'unstable') and version is None:
+            raise MesonException(f'{xml_state} protocols require a version number.')
+
+        if self.protocols_dep is None:
+            self.protocols_dep = self.interpreter.func_dependency(state.current_node, ['wayland-protocols'], {})
+
+        if self.pkgdatadir is None:
+            self.pkgdatadir = self.protocols_dep.get_variable(pkgconfig='pkgdatadir', internal='pkgdatadir')
+
+        if xml_state == 'stable':
+            if version is not None:
+                mlog.warning('stable protocols do not require a version number.', location=self.interpreter.current_node, once=True)
```

I would prefer a hard error.

---

Also this validation check could be moved up above to where we check and error out on a missing one for unstable/staging, as it's better to error out or warn on invalid arguments earlier rather than later, and currently it happens after checking the environment state (dependency lookup).